### PR TITLE
tests: mem protection test check CPU scrubs registers after syscall

### PR DESCRIPTION
**Pull request by maksimmasalski**
Add new memory protection test, that verifies CPU scrubs registers after syscall. Leaving of sensitive data in the CPU registers, can make open a door for attacks and vulnerabilities.

To verify that CPU registers are scrubbed, I created test test_after_syscall_cpu_scrubs_regs()
Test verifies upon exit of a system call back to the calling thread, the kernel scrubs CPU registers for sensitive data. 

_Please review and give comments, if you think that my solution is not correct. Also I didn't implement that test for ARC, because not an expert in ARC, so looking for community help who knows ARC.
ARC currently not implemented and not supported for that test._

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

### CODE REVIEW

**Comment by maksimmasalski** [comment]
@andrewboie I'm looking for your comments about that PR, I need someone who is good at ASM and CPU registers to review that code suggestion.

File: tests/kernel/mem_protect/syscalls/src/main.c
Comment (sjg20) [inline tests/kernel/mem_protect/syscalls/src/main.c:483]: How about just #else

File: tests/kernel/mem_protect/syscalls/src/main.c
Comment (sjg20) [inline tests/kernel/mem_protect/syscalls/src/main.c:57]: You could use an intermediate to make use of DB_VAL here

**Review by sjg20 — APPROVED** [review]

File: tests/kernel/mem_protect/syscalls/src/main.c
Comment (maksimmasalski) [reply tests/kernel/mem_protect/syscalls/src/main.c:58]: I can't find out how to use C defined variable in asm code. [DB_VAL] is not working. Do you know how to do that?

File: tests/kernel/mem_protect/syscalls/src/main.c
Comment (maksimmasalski) [reply tests/kernel/mem_protect/syscalls/src/main.c:485]: I'm afraid if to use there #else, branches below will be broken.

**Comment by maksimmasalski** [comment]
@LeiW000 @wentongwu Please review that test when you have time.

**Review by LeiW000 — APPROVED** [review]

**Review by wentongwu — APPROVED** [review]

**Comment by ioannisg** [comment]
> Also I didn't implement that test for ARC, because not an expert in ARC, so looking for community help who knows ARC.
ARC currently not implemented and not supported for that test.

@maksimmasalski IMHO this test is validating architecture-specific low-level implementation of the syscall framework and therefore it could better fit under the ARCH-specific test directory. I think that the mem_protect/syscalls suite should not deal with ARCH-related assembly and, instead, it should verify the syscall stuff up to the kernel level.

There's already a syscall-related test under tests/arch/arm/arm_thread_swap, which you could use for Cortex-M register scrub, and the same thing should be more or less done for each Architecture that supports system calls (user mode). Otherwise this is not scaling well, considering that we're soon getting more ARCHes doing user mode (Risc-v, other ARM variants, etc.); we do not want to maintain all the different ASM code for all ARCHes inside a kernel test.

As an ARCH-maintainer, I would like all ARCH-related code validation with ARCH-specific assembly considerations to be under a single test directory.

This would also solve your problem of not doing the work for ARC; essentially it is the ARC-maintainer's responsibility to test the underlying assembly stuff around system calls.

**Comment by maksimmasalski** [comment]
> @maksimmasalski IMHO this test is validating architecture-specific low-level implementation of the syscall framework and therefore it could better fit under the ARCH-specific test directory. I think that the mem_protect/syscalls suite should not deal with ARCH-related assembly and, instead, it should verify the syscall stuff up to the kernel level.
> 
> There's already a syscall-related test under tests/arch/arm/arm_thread_swap, which you could use for Cortex-M register scrub, and the same thing should be more or less done for each Architecture that supports system calls (user mode). Otherwise this is not scaling well, considering that we're soon getting more ARCHes doing user mode (Risc-v, other ARM variants, etc.); we do not want to maintain all the different ASM code for all ARCHes inside a kernel test.
> 
> As an ARCH-maintainer, I would like all ARCH-related code validation with ARCH-specific assembly considerations to be under a single test directory.
> 
> This would also solve your problem of not doing the work for ARC; essentially it is the ARC-maintainer's responsibility to test the underlying assembly stuff around system calls.

Thank you for the reply. I understand key points:
1. Cortex-M has tests/arch/arm/arm_thread_swap where I can put my test code
2. The same I can do for x86, put my test code into tests/arch/x86
3. If I want to add ARC test, I can ask you or ask you for your help, if I will create a draft idea of that test.

Do I have that right?

**Comment by ioannisg** [comment]
> Cortex-M has tests/arch/arm/arm_thread_swap where I can put my test code

Yeap, that's right.
This makes sense, cause people reading the ARM tests understand better what r0 - r3 registers are, but I cannot say the same for generic readers of kernel tests :)

> The same I can do for x86, put my test code into tests/arch/x86

Yes, if you wish to, but I won't be able to review those.

> If I want to add ARC test, I can ask you or ask you for your help, if I will create a draft idea of that test.

Better to ask anyone fro Synopsys about this.

**Review by ioannisg — CHANGES_REQUESTED** [review]
Let's not merge this PR as is (at least this is my opinion) and instead provide ARCH-specific tests in the respective arch-specific test-suites)

**Comment by maksimmasalski** [comment]
Started to move test to arch folder, and refactored it by removing unnecessary operations in the code.
Started from ARM and created new PR #27878

**Pull request closed without merging**
